File: php/JsonStorage.php
<?php

declare(strict_types=1);

/**
 * JSON Storage utility class for payment methods
 */
class JsonStorage
{
    private static string $dataDir = __DIR__ . '/data';
    private static string $paymentMethodsFile = 'payment_methods.json';

    /**
     * Initialize storage directory
     */
    public static function init(): void
    {
        if (!is_dir(self::$dataDir)) {
            mkdir(self::$dataDir, 0755, true);
        }
    }

    /**
     * Generate unique ID for payment methods
     */
    public static function generateId(): string
    {
        return 'pm_' . uniqid() . '_' . bin2hex(random_bytes(8));
    }

    /**
     * Read all payment methods
     */
    public static function readPaymentMethods(): array
    {
        $filePath = self::$dataDir . '/' . self::$paymentMethodsFile;
        
        if (!file_exists($filePath)) {
            return [];
        }
        
        $content = file_get_contents($filePath);
        if ($content === false) {
            return [];
        }
        
        $data = json_decode($content, true);
        return $data ?? [];
    }

    /**
     * Write payment methods to storage
     */
    public static function writePaymentMethods(array $methods): bool
    {
        self::init();
        $filePath = self::$dataDir . '/' . self::$paymentMethodsFile;
        
        $json = json_encode($methods, JSON_PRETTY_PRINT);
        return file_put_contents($filePath, $json) !== false;
    }

    /**
     * Add a new payment method
     */
    public static function addPaymentMethod(array $paymentMethod): bool
    {
        $paymentMethod['createdAt'] = date('c');
        $paymentMethod['updatedAt'] = date('c');
        
        $methods = self::readPaymentMethods();
        $methods[] = $paymentMethod;
        
        return self::writePaymentMethods($methods);
    }

    /**
     * Find payment method by ID
     */
    public static function findPaymentMethod(string $id): ?array
    {
        $methods = self::readPaymentMethods();
        
        foreach ($methods as $method) {
            if ($method['id'] === $id) {
                return $method;
            }
        }
        
        return null;
    }

    /**
     * Update an existing payment method
     */
    public static function updatePaymentMethod(string $id, array $updateData): bool
    {
        $methods = self::readPaymentMethods();
        $updated = false;
        
        foreach ($methods as $index => $method) {
            if ($method['id'] === $id) {
                // Update only allowed fields
                if (isset($updateData['nickname'])) {
                    $methods[$index]['nickname'] = $updateData['nickname'];
                }
                
                if (isset($updateData['isDefault'])) {
                    // If setting this as default, unset default from all others
                    if ($updateData['isDefault']) {
                        foreach ($methods as $i => $m) {
                            $methods[$i]['isDefault'] = false;
                        }
                    }
                    $methods[$index]['isDefault'] = $updateData['isDefault'];
                }
                
                // Update timestamp
                $methods[$index]['updatedAt'] = date('c');
                $updated = true;
                break;
            }
        }
        
        if ($updated) {
            return self::writePaymentMethods($methods);
        }
        
        return false;
    }

    /**
     * Check if a payment method exists by ID
     */
    public static function paymentMethodExists(string $id): bool
    {
        return self::findPaymentMethod($id) !== null;
    }

    /**
     * Validate payment method data
     */
    public static function validatePaymentMethod(array $data): array
    {
        $errors = [];
        
        if (empty($data['cardBrand'])) {
            $errors[] = 'Card brand is required';
        }
        
        if (empty($data['last4']) || strlen($data['last4']) !== 4) {
            $errors[] = 'Valid last 4 digits are required';
        }
        
        if (empty($data['expiryMonth']) || !is_numeric($data['expiryMonth']) || 
            $data['expiryMonth'] < 1 || $data['expiryMonth'] > 12) {
            $errors[] = 'Valid expiry month is required';
        }
        
        if (empty($data['expiryYear']) || !is_numeric($data['expiryYear'])) {
            $errors[] = 'Valid expiry year is required';
        }
        
        return $errors;
    }

    /**
     * Validate update data for payment method editing
     */
    public static function validateUpdateData(array $data): array
    {
        $errors = [];
        
        // Nickname validation (optional)
        if (isset($data['nickname']) && strlen($data['nickname']) > 100) {
            $errors[] = 'Nickname must be 100 characters or less';
        }
        
        // isDefault validation (optional but must be boolean if provided)
        if (isset($data['isDefault']) && !is_bool($data['isDefault'])) {
            $errors[] = 'isDefault must be a boolean value';
        }
        
        return $errors;
    }
}

File: dotnet/JsonStorage.cs
using System.Text.Json;

namespace CardPaymentSample;

/// <summary>
/// Simple JSON-based data storage for payment methods
/// </summary>
public static class JsonStorage
{
    private const string DataDir = "data";
    private const string PaymentMethodsFile = "data/payment_methods.json";

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = true
    };

    /// <summary>
    /// Ensure data directory exists
    /// </summary>
    private static void EnsureDataDir()
    {
        if (!Directory.Exists(DataDir))
        {
            Directory.CreateDirectory(DataDir);
        }
    }

    /// <summary>
    /// Load payment methods from JSON file
    /// </summary>
    public static async Task<List<PaymentMethod>> LoadPaymentMethodsAsync()
    {
        EnsureDataDir();

        if (!File.Exists(PaymentMethodsFile))
        {
            return new List<PaymentMethod>();
        }

        try
        {
            var json = await File.ReadAllTextAsync(PaymentMethodsFile);
            return JsonSerializer.Deserialize<List<PaymentMethod>>(json, JsonOptions) ?? new List<PaymentMethod>();
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error loading payment methods: {ex.Message}");
            return new List<PaymentMethod>();
        }
    }

    /// <summary>
    /// Save payment methods to JSON file
    /// </summary>
    public static async Task SavePaymentMethodsAsync(List<PaymentMethod> methods)
    {
        EnsureDataDir();

        try
        {
            var json = JsonSerializer.Serialize(methods, JsonOptions);
            await File.WriteAllTextAsync(PaymentMethodsFile, json);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error saving payment methods: {ex.Message}");
        }
    }

    /// <summary>
    /// Add a new payment method
    /// </summary>
    public static async Task<PaymentMethod> AddPaymentMethodAsync(StoredPaymentMethodData data)
    {
        var methods = await LoadPaymentMethodsAsync();

        // Generate unique ID
        var id = $"pm_{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}_{Guid.NewGuid().ToString()[^9..]}";

        // Create payment method object
        var paymentMethod = new PaymentMethod
        {
            Id = id,
            StoredPaymentToken = data.StoredPaymentToken,
            CardBrand = data.CardBrand,
            Last4 = data.Last4,
            ExpiryMonth = data.ExpiryMonth,
            ExpiryYear = data.ExpiryYear,
            Nickname = data.Nickname,
            IsDefault = data.IsDefault,
            CustomerData = data.CustomerData,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        // Handle default payment method
        if (data.IsDefault)
        {
            // Remove default from all existing methods
            foreach (var method in methods)
            {
                method.IsDefault = false;
            }
        }
        else if (methods.Count == 0)
        {
            // Make first method default
            paymentMethod.IsDefault = true;
        }

        methods.Add(paymentMethod);
        await SavePaymentMethodsAsync(methods);

        return paymentMethod;
    }

    /// <summary>
    /// Find payment method by ID
    /// </summary>
    public static async Task<PaymentMethod?> FindPaymentMethodAsync(string id)
    {
        var methods = await LoadPaymentMethodsAsync();
        return methods.FirstOrDefault(m => m.Id == id);
    }

    /// <summary>
    /// Get all payment methods formatted for display
    /// </summary>
    public static async Task<List<FormattedPaymentMethod>> GetFormattedPaymentMethodsAsync()
    {
        var methods = await LoadPaymentMethodsAsync();

        return methods.Select(method => new FormattedPaymentMethod
        {
            Id = method.Id,
            Brand = method.CardBrand,
            Last4 = method.Last4,
            Expiry = $"{method.ExpiryMonth}/{method.ExpiryYear}",
            Nickname = method.Nickname,
            IsDefault = method.IsDefault
        }).ToList();
    }

    /// <summary>
    /// Update an existing payment method (nickname and default status only)
    /// </summary>
    public static async Task<PaymentMethod?> UpdatePaymentMethodAsync(string id, PaymentMethodEditData editData)
    {
        var methods = await LoadPaymentMethodsAsync();
        var method = methods.FirstOrDefault(m => m.Id == id);
        
        if (method == null)
        {
            return null;
        }

        // Update nickname
        method.Nickname = editData.Nickname;
        method.UpdatedAt = DateTime.UtcNow;

        // Handle default payment method change
        if (editData.IsDefault && !method.IsDefault)
        {
            // Remove default from all existing methods
            foreach (var m in methods)
            {
                m.IsDefault = false;
            }
            method.IsDefault = true;
        }
        else if (!editData.IsDefault && method.IsDefault)
        {
            method.IsDefault = false;
        }

        await SavePaymentMethodsAsync(methods);
        return method;
    }

    /// <summary>
    /// Set a payment method as default (and remove default from others)
    /// </summary>
    public static async Task<bool> SetDefaultPaymentMethodAsync(string id)
    {
        var methods = await LoadPaymentMethodsAsync();
        var targetMethod = methods.FirstOrDefault(m => m.Id == id);
        
        if (targetMethod == null)
        {
            return false;
        }

        // Remove default from all methods
        foreach (var method in methods)
        {
            method.IsDefault = false;
        }

        // Set target method as default
        targetMethod.IsDefault = true;
        targetMethod.UpdatedAt = DateTime.UtcNow;

        await SavePaymentMethodsAsync(methods);
        return true;
    }
}